### PR TITLE
feat:  add `scalars` config option for custom GraphQL [scalar -> casual] mappings

### LINKED ISSUE #20: Feature Request: Custom Scalar Definitions
@ardeois to begin with, thanks for all the work being put in this plugin.

In our app we use [AWS Appsync Scalars](https://docs.aws.amazon.com/appsync/latest/devguide/scalars.html), which add specific GraphQL types for dates. Unfortunately, one of them is `AWSTimeStamp` which maps to a timestamp (`number`).

Currently, as it stands, this plugin translates all the custom scalars into strings which breaks our TS validation, since we expected a `number`. 

Given the fact that you said that [custom scalars need to accept custom definitions](https://github.com/ardeois/graphql-codegen-typescript-mock-data/blob/1e292c8bc3237a06603e21f910fae545e43e4a83/src/index.ts#L104-L105), my question would be whether you have immediate plans to support it  and/or whether you would be open to accepting PRs to add this functionality.

Thanks in advance

### PR DESCRIPTION
**Pull request by 3nvi**
## Background

Some apps use custom scalar for date-related stuff. This caused issues with the way `graphql-codegen-typescript-mock-data` handles autogenerated data, since a "random string" (the current  default), may not always be what you need.

This PR allows you to define mapping from your custom scalars to a casual embedded generator, allowing you to specify exactly how your scalar's autogenerated value will be populated

Closes #20 

## Caveats
To make configuration simple, a few "shortcuts" were taken. Namely:

- the mapping only accepts a `casual` key (function or string) as a valid autogenerated value in order to align with the rest of the library
- if a function is passed, then it will be called with its default arguments, since providing custom parameters for a function through a YAML file, would complicate things more than normal

### CODE REVIEW

**Comment by 3nvi** [comment]
There is a minorr issue with stringification that tests didn't catch. Will update the PR

File: src/index.ts [hunk 1]
```diff
@@ -101,9 +102,16 @@ const getNamedType = (
...
-                        // This could be improved with a custom scalar definition in the config
-                        return `'${casual.word}'`;
+                        // it's a scalar, let's use a string as a value if there is no custom
+                        // mapping for this particular scalar
+                        if (!customScalars || !customScalars[foundType.name]) {
+                            return `'${casual.word}'`;
+                        }
+
+                        // If there is a mapping to a `casual` type, then use this
+                        const embeddedGenerator = casual[customScalars[foundType.name]];
```
Comment (ardeois) [inline src/index.ts:112]: What if there is no mapping? This would return `undefined`
Maybe throw an exception when there is no mapping?

A future improvement would be to support custom functions from the config, but I think this is a good first step.
We'll see if people need custom functions in a future issue/PR

File: tests/typescript-mock-data.spec.ts [hunk 2]
```diff
@@ -168,3 +168,12 @@ it('should generate mock data with as-is types and enums if typenames is "keep"'
     expect(result).not.toMatch(/ABC(TYPE|STATUS)/);
     expect(result).toMatchSnapshot();
 });
+
+it('should generate the `casual` data for a particular scalar mapping', async () => {
+    const result = await plugin(testSchema, [], { scalars: { AnyObject: 'email' } });
+
+    const emailRegex = /(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))/;
+    expect(result).toBeDefined();
+    expect(emailRegex.test(result as string)).toBeTruthy();
```
Comment (ardeois) [inline tests/typescript-mock-data.spec.ts:177]: Our test suite force `casual` to return a predictable result, so I think you can check the `result` string directly instead of using a regex

File: tests/typescript-mock-data.spec.ts [hunk 3]
```diff
@@ -168,3 +168,12 @@ it('should generate mock data with as-is types and enums if typenames is "keep"'
...
+    expect(result).toMatchSnapshot();
+});
```
Comment (ardeois) [inline tests/typescript-mock-data.spec.ts:179]: Could you add a test for custom scalar that is not a string?

File: src/index.ts (hunk 1 above)
Comment (3nvi) [reply src/index.ts:112]: If you scroll up a bit above, you will see

```tsx

  // it's a scalar, let's use a string as a value if there is no custom
   // mapping for this particular scalar
   if (!customScalars || !customScalars[foundType.name]) {
      return `'${casual.word}'`;
    }
```

so I'm defaulting to a word if there is no scalar. Wouldn't that suffice?

File: tests/typescript-mock-data.spec.ts (hunk 3 above)
Comment (3nvi) [reply tests/typescript-mock-data.spec.ts:179]: sure thing. Will update soon and ping

File: tests/typescript-mock-data.spec.ts (hunk 2 above)
Comment (3nvi) [reply tests/typescript-mock-data.spec.ts:177]: oh cool, didn't dig in too much and with a quick pass, I wasn't able to find something. I'll make sure to update

File: tests/typescript-mock-data.spec.ts (hunk 3 above)
Comment (ardeois) [reply tests/typescript-mock-data.spec.ts:186]: Thanks for the second test, but it seems like both test generate a string
https://github.com/ardeois/graphql-codegen-typescript-mock-data/pull/21/files/a3511d4a629bedb2a231dc74197ee166e315ad1c..b43a92e5b172f8d035419270cbc65acf80dd12cb#diff-ce1f2f758ad7b0b108a11d9a0171e13dR34

Shouldn't you test with a number?

File: src/index.ts [hunk 4]
```diff
@@ -101,9 +102,24 @@ const getNamedType = (
...
+                        // If there is a mapping to a `casual` type, then use it and make sure
+                        // to call it if it's a function
+                        const embeddedGenerator = casual[customScalars[foundType.name]];
+                        const value = typeof embeddedGenerator === 'function' ? embeddedGenerator() : embeddedGenerator;
+
+                        if (typeof value === 'string') {
+                            return `'${value}'`;
+                        }
+                        if (typeof value === 'object') {
+                            return `${JSON.stringify(value)}`;
```
Comment (3nvi) [inline src/index.ts:120]: @ardeois added `JSON.stringify` to handle autogenerated types of array/object.

Those will not normally appear since this handler has to do with a particular Scalar (and never an array of a Scalar), but still nothing stops the users from specifying `casual.word_list` which would cause issues.

All in all, this handles an edge case where user input is not following the schema. If you feel it's unneeded, I can remove it

File: tests/typescript-mock-data.spec.ts (hunk 3 above)
Comment (3nvi) [reply tests/typescript-mock-data.spec.ts:186]: Yeah updated it to something a bit more extreme just to be safe
